File: CoUML-app/CoUML TS/Attribute.ts

export enum VisabilityType{
    Private="-",
    Public="+",
    Protected="#",
    Package="~",
    LocalScope=""
}

/*
export enum DataType{
    Int,
    Double,
    Float,
    String
}
*/

//maybe add a value for type?
export abstract class DataType{

}


export class Attribute{
    visibility:VisabilityType;
    name:string;
    type:DataType;//not sure if we want enums we can make an abstract rn
    multiplicity:number = null;//diagram says int[10..-1]=null idk. no car in ts. make * -1
    default:string = null;//says =null
    propertyString:string = null;//says =null
}


export interface AttributeCollectionIF{
    Iterator():AttributeIteratorIF;
}

//after that rewrite in c# for both ends
//collections are part of iterator pattern
export class AttributeCollection implements AttributeCollectionIF{
    attributes: Attribute[];
    
    //creates an iteratorIF
    Iterator():AttributeIteratorIF{
        return new AttributeIterator();
    }
}

export interface AttributeIteratorIF{
    hasNextAttribute():boolean;
    getNextAttribute():Attribute;
    hasPreviousAttribute():boolean;
    getPreviousAttribute():Attribute;
}

//fetches objects from collection
export class AttributeIterator implements AttributeIteratorIF{
    collection:AttributeCollection;
    position:number = 0;

    hasNextAttribute():boolean{
        return this.collection.attributes.length > this.position+1;
    }
    getNextAttribute():Attribute{
        if(this.hasNextAttribute){
            return this.collection.attributes[++this.position];
        }
    }
    hasPreviousAttribute():boolean{
        return this.position-1 > 0;
    }
    getPreviousAttribute():Attribute{
        if(this.hasPreviousAttribute){
            return this.collection.attributes[--this.position];
        }
    }
}

File: CoUML-app/Models/Attribute.cs

enum VisabilityType{//enums cant be strings in c#
    Private=1,//-
    Public=2,//+
    Protected=3,//#
    Package=4,//~
    LocalScope=5//
}

abstract class Datatype{

}


class Attribute{
    VisabilityType visability;
    string name;
    Datatype type;
    int multiplicity;
    string defaultName = null;//default value in ts
    string propertyString = null;
}

interface AttributeCollectionIF{
    AttributeIteratorIF Iterator();
}

class AttributeCollection : AttributeCollectionIF{
    public Attribute[] attributes;//needs to be a dynamic array

    AttributeIteratorIF AttributeCollectionIF.Iterator(){
        return new AttributeIterator();
    }
}

interface AttributeIteratorIF{
    bool hasNextAttribute();
    Attribute getNextAttribute();
    bool hasPreviousAttribute();
    Attribute getPreviousAttribute();
}

class AttributeIterator : AttributeIteratorIF{
    AttributeCollection collection;
    int position = 0;


    bool AttributeIteratorIF.hasNextAttribute(){
        return this.collection.attributes.Length > this.position+1;
    }
    Attribute AttributeIteratorIF.getNextAttribute(){
        AttributeIteratorIF iterator = this;//is there another way for this to work? i couldnt get it to work
        if (iterator.hasNextAttribute()){
            return this.collection.attributes[++this.position];
        }
        return null;
    }
    bool AttributeIteratorIF.hasPreviousAttribute(){
        return this.position-1 > 0;
    }
    Attribute AttributeIteratorIF.getPreviousAttribute(){
        AttributeIteratorIF iterator = this;
        if(iterator.hasPreviousAttribute()){
            return this.collection.attributes[--this.position];
        }
        return null;
    }
}